modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericComparisonTypeCoercionTest.java

```diff
@@ -0,0 +1,965 @@
+
+                forTypePair(NumericPair.DECIMAL_3_1_REAL)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(castTo(NativeTypes.DOUBLE)),
+
+                forTypePair(NumericPair.DECIMAL_3_1_DOUBLE)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(ofTypeWithoutCast(NativeTypes.DOUBLE)),
+
+
+                forTypePair(NumericPair.DECIMAL_5_3_DECIMAL_5_3)
+                        .firstOpMatches(ofTypeWithoutCast(Types.DECIMAL_5_3))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_5_3)),
+
+                forTypePair(NumericPair.DECIMAL_5_3_DECIMAL_5_0)
+                        .firstOpMatches(castTo(Types.DECIMAL_8_3))
+                        .secondOpMatches(castTo(Types.DECIMAL_8_3)),
+
+                forTypePair(NumericPair.DECIMAL_5_3_DECIMAL_6_1)
+                        .firstOpMatches(castTo(Types.DECIMAL_8_3))
+                        .secondOpMatches(castTo(Types.DECIMAL_8_3)),
+
+                forTypePair(NumericPair.DECIMAL_5_3_DECIMAL_8_3)
+                        .firstOpMatches(castTo(Types.DECIMAL_8_3))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_8_3)),
+
+                forTypePair(NumericPair.DECIMAL_5_3_REAL)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(castTo(NativeTypes.DOUBLE)),
+
+                forTypePair(NumericPair.DECIMAL_5_3_DOUBLE)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(ofTypeWithoutCast(NativeTypes.DOUBLE)),
+
+
+                forTypePair(NumericPair.DECIMAL_5_0_DECIMAL_5_0)
+                        .firstOpMatches(ofTypeWithoutCast(Types.DECIMAL_5_0))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_5_0)),
+
+                forTypePair(NumericPair.DECIMAL_5_0_DECIMAL_6_1)
+                        .firstOpMatches(castTo(Types.DECIMAL_6_1))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_6_1)),
+
+                forTypePair(NumericPair.DECIMAL_5_0_DECIMAL_8_3)
+                        .firstOpMatches(castTo(Types.DECIMAL_8_3))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_8_3)),
+
+                forTypePair(NumericPair.DECIMAL_5_0_REAL)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(castTo(NativeTypes.DOUBLE)),
+
+                forTypePair(NumericPair.DECIMAL_5_0_DOUBLE)
+                        .firstOpMatches(castTo(NativeTypes.DOUBLE))
+                        .secondOpMatches(ofTypeWithoutCast(NativeTypes.DOUBLE)),
+
+
+                forTypePair(NumericPair.DECIMAL_6_1_DECIMAL_6_1)
+                        .firstOpMatches(ofTypeWithoutCast(Types.DECIMAL_6_1))
+                        .secondOpMatches(ofTypeWithoutCast(Types.DECIMAL_6_1)),
```

Comment:
This looks similar to NumericPair.DECIMAL_8_3_DECIMAL_8_3 and NumericPair.DECIMAL_5_3_DECIMAL_5_3.

Reply:
that's true, but I would prefer to leave this as is. 

To wrote all those test cases, first, I defined a list of types that I would like to verify. Among others, I definitely did want to verify different DECIMAL types: exactly the same types, types with equal precision but different scale, types with equal scale but different precision, completely different types. Then, I made a Cartesian join of the type list a leave only those pairs where first type is lesser then or equals to second. That's how I end up with *duplicates. Technically, it's possible to exclude such pairs, but it much easier and  safer just to include them as valid test cases. After all, every test case takes couple of milliseconds only. Not a big deal, I believe.